### ISSUE TITLE
🐛 Colored output from standard grep is not parsed correctly

### ISSUE THREAD

**Issue by dandavison** [dandavison/delta#794]
Delta officially supports two ways of highlighting grep output:

1. `git grep` with delta configured as git's pager either via `core.pager = delta` or `pager.grep = delta`
2. `rg --json ... | delta`

However, delta does try to highlight other grep output that is piped to it. This issue shows it failing to handle output from standard grep (`grep (BSD grep) 2.5.1-FreeBSD`)

<table><tr><td><img width=600px src="https://user-images.githubusercontent.com/52205/143273008-648b3fe7-73e6-4ff3-b9e8-5aadaf69314f.png" alt="image" /></td></tr></table>

<table><tr><td><img width=600px src="https://user-images.githubusercontent.com/52205/143272159-29254e0a-cf89-4e71-9628-79a48aa6c969.png" alt="image" /></td></tr></table>

cf

Similarly

<table><tr><td><img width=600px src="https://user-images.githubusercontent.com/52205/143273924-be49b509-cecf-4a31-ae17-5a7e59a3a152.png" alt="image" /></td></tr></table>

**Comment by dandavison** [comment]
It looks like `parse_style_sections` is failing to parse the ANSI escape sequences emitted by BSD grep and GNU grep. For example,

```
delta(794-ansi-parse-bug *) grep --color=always -Hn s src/**/*.rs | head | bat -A   
src/align.rs:1:u␛[01;31m␛[Ks␛[m␛[Ke·␛[01;31m␛[Ks␛[m␛[Ktd::cmp::max;␊
src/align.rs:2:u␛[01;31m␛[Ks␛[m␛[Ke·␛[01;31m␛[Ks␛[m␛[Ktd::collection␛[01;31m␛[Ks␛[m␛[K::VecDeque;␊
src/align.rs:4:con␛[01;31m␛[Ks␛[m␛[Kt·SUBSTITUTION_COST:·u␛[01;31m␛[Ks␛[m␛[Kize·=·1;␊
src/align.rs:5:con␛[01;31m␛[Ks␛[m␛[Kt·DELETION_COST:·u␛[01;31m␛[Ks␛[m␛[Kize·=·1;␊
src/align.rs:6:con␛[01;31m␛[Ks␛[m␛[Kt·INSERTION_COST:·u␛[01;31m␛[Ks␛[m␛[Kize·=·1;␊
src/align.rs:11:····Sub␛[01;31m␛[Ks␛[m␛[Ktitution,␊
src/align.rs:13:····In␛[01;31m␛[Ks␛[m␛[Kertion,␊
src/align.rs:16:u␛[01;31m␛[Ks␛[m␛[Ke·Operation::*;␊
src/align.rs:18:///·Needleman-Wun␛[01;31m␛[Ks␛[m␛[Kch·/·Wagner-Fi␛[01;31m␛[Ks␛[m␛[Kcher·table·for·computation·of·edit·di␛[01;31m␛[Ks␛[m␛[Ktance·and·a␛[01;31m␛[Ks␛[m␛[K␛[01;31m␛[Ks␛[m␛[Kociated␊
src/align.rs:21:␛[01;31m␛[Ks␛[m␛[Ktruct·Cell·{␊
```

```
delta(794-ansi-parse-bug *) grep --color=always -Hn s src/**/*.rs | delta        
[src/ansi/mod.rs:76] &s[start..end] = "u"
[src/ansi/mod.rs:76] &s[start..end] = "\u{1b}"
[src/ansi/mod.rs:76] &s[start..end] = "\u{1b}["
[src/ansi/mod.rs:76] &s[start..end] = "\u{1b}"
[src/ansi/mod.rs:76] &s[start..end] = "\u{1b}[Ktd::cmp::max;"
thread 'main' panicked at 'String mismatch encountered while superimposing style sections: 's 't'', src/paint.rs:921:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
src/align.rs:1:
```

The rust `dbg!` output shows that some ANSI escape sequences are being parsed as text.

**Comment by alerque** [comment]
I just started getting this panic when used as git's diff filter.

**Comment by dandavison** [comment]
Hi @alerque, thanks for this. Could you post the raw diff that's causing this? I.e. the file produced by`git diff (your rags) > diff.txt`.

**Comment by alerque** [comment]
I can't post a dump of the full content of the diff as it has some private unpublished stuff, but I have at least two commits in this project now that are giving the same panic results so when I get a second I'll try to comb through them and figure out what bit is delta is tripping on.

**Comment by alerque** [comment]
It isn't *just* the raw content.

```console
$ git show 57e5082
snip... thread 'main' panicked at 'String mismatch encountered while superimposing style sections: '↵' vs '↴'', src/paint.rs:921:17
$ git show 57e5082 > diff.txt
$ cat diff.txt | delta
snip... correctly rendered diff of 57e5082
```

**Comment by alerque** [comment]
The difference above is in the delta arguments used. In my Git config I have:

```ini
[delta]
	features = side-by-side line-numbers decorations line-wrap
	hunk-header-style = file line-number syntax
    whitespace-error-style = 22 reverse
	syntax-theme = Monokai Extended
	max-line-length = 0
```

If I drop the side-by-side feature it works fine as a git pager. Only if I add that as an argument and have it render the diff.txt as per my example that works fine too.

**Comment by dandavison** [comment]
Do you have git configured to display moved chunks with `color-moved`? Is the problem unicode characters in moved chunks?

If you could try to make the diff smaller so that it still gives the problem but isn't too private to share that would help a lot!

**Comment by alerque** [comment]
If I could get the output diff content to make delta panic it would be easy to break the diff up into chunks and find offending content and sent a version that doesn't publicize anything private. The issue is I haven't gotten it to fail except when used directly as a git's diff viewer.

Yes I do have `git -c diff.colorMoved=zebra` set and yes it does work (including in side-by-side mode) if I disable that. So this is some combination of specific content + delta.side-by-side + git.diff.colorMoved.

**Comment by dandavison** [comment]
> The issue is I haven't gotten it to fail except when used directly as a git's diff viewer.

Oh I'm sorry, I should have pointed out -- you need to use `git show --color=always` to force git to emit ANSI color escape sequences when its output is not going to the screen.

**Comment by alerque** [comment]
Okay I have something for you. I exctracted Git's output from the offending commit like this:

```console
$ git -c pager.show=nvim show 57e5082
```

In Neovim I ran `:w diff-codes.txt` to get a file with all the original escape codes intact.

Then tried to render it like this:

```console
$ < diff-codes.txt delta --side-by-side --pager cat
```

Seeing that that produced the same panic, I then went through and started deleting lines until I found the offending one. Here is a file to replicate the problem: 

Delta appears to have a beef with a lite that has a very much not-private previously published line of Bible text taken from Hebrews 4:5 (ESV):

> And again in this passage he said, "They shall not enter my rest."

This (and a few hundred surrounding lines) were deleted from one file and moved to another in this commit. In messing around with the content, the problem went away if I changed the part before the comma to be "foo".

Lastly in preparing this post I had a moment of panic because I couldn't replicate it on a second pass when I was all done with the writeup. The difference turned out to be terminal width: I get a panic on a narrow terminal (98 cols) but not a wide one (197 cols).

**Comment by dandavison** [comment]
Great, thanks! I can reproduce that (as you say, dependent on terminal width).

I also found an example while you were doing that. Going backwards through nginx commits with `side-by-side` active causes the error below, with a wide terminal and side-by-side.
 
```
nginx(master) git log --reverse -p | delta --side-by-side > /dev/null
[src/paint.rs:925] s1 = "ex_module ��� ������ ������� �\u{363}� "
[src/paint.rs:926] s2 = "ex_module ��� ������ ������� �\u{363}�↵"
thread 'main' panicked at 'String mismatch encountered while superimposing style sections: ' ' vs '↵'', src/paint.rs:927:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by dandavison** [comment]
Fixed by #976
